Add a pass to simplify single-character choices

markw65 opened the pull request:
I have a modified version of examples/xml.peggy in my project, and was trying to optimize it for speed. I found a few places where it could be rewritten to avoid backtracking, getting about a 20% win, but with some readability cost.

Then I noticed that there are a number of long lists of alternate character classes, and they end up being implemented very inefficiently; effectively it tries one regex after another in sequence.

I manually combined all the character classes for BaseChar into a single character class, and got a nearly 2x speed up.

So I've added a pass that does that automatically. The result is a more than 3x speedup parsing the xml files that I work with (which seem fairly typical).

The pass takes a rule like `chars: "a" / "b" / [d-r]` and turns it into `chars: [abd-r]`

markw65 commented:
I've put it under an option, `mergeCharacterClasses` because enabling it by default breaks a number of tests, which have specific expectations for error messages:

```
    Expected value to strictly be equal to:
      "Expected \"a\", \"b\", or \"c\" but \"d\" found."
    Received:
      "Expected [abc] but \"d\" found."
```

Since that test is specifically designed to ensure that an appropriate error message is output for `choice` nodes I didn't want to just change the expected output. But I'd be happy to change the grammar for such tests so that the choice nodes remain as choice nodes, and enable the optimization by default if that would be preferable.

Mingun reviewed (changes requested):
I generally agree, that such optimization is very useful, but we need more tests to ensure, that it works as expected and does not processes pieces which it shouldn't.

Can you please add tests for negative scenarios, at least.

I'm not sure, should we add a setting for this optimization and should it be enabled by default. It seems to me, that settings is superfluous and we can apply this optimization always. So yes, I'm fine with changing expected error messages in failed tests

markw65 commented on test/unit/compiler/passes/merge-character-classes.spec.js:
```diff
@@ -0,0 +1,38 @@
+  it("Merges single character literals, class, and ref_rules", () => {
+    expect(pass).to.changeAST(
+      [
+        "one = three / 'a' / [c-g]",
+        "two = three / 'P' / [Q-T]",
+        "three = 'x' / [u-w]",
```
Mostly agreed - I'll certainly add more tests.

> For example, if three rule will contain $ operator, or action block

In that case, its `expression`'s  type will be `text` or `action` and `asClass` will ignore it. So I don't need to do anything to handle that.

But you made me realize that I should explicitly check for `text` and look through it, because we actually want 
`foo = "a" / $("b" / [xyz]) / "X"` to get converted to `foo = [Xabxyz]`.

markw65 commented:
I just realized I need to update `lib/peggy.js` and `docs/js/examples.js` now that this is done unconditionally. Should I do that now, or wait until you're happy with the state of the commit?

hildjj commented:
This looks good to me, I think.  @Mingun any other comments?

hildjj commented:
Except one needed addition:  a CHANGELOG entry, please.

markw65 commented:
> Except one needed addition: a CHANGELOG entry, please

I put it in "minor" updates - let me know if if should be somewhere else...

hildjj commented:
Sorry for moving the goalposts, but once I thought of coverage on #427, there are a few lines that need to be hit on this one in lib/compiler/passes/merge-character-classes.js, unless they are particularly hard to reach.

Mingun commented on lib/compiler/passes/merge-character-classes.js:
```diff
@@ -0,0 +1,144 @@
+        return asClass(node.expression);
+      }
+      if (node.type === "literal" && node.value.length === 1) {
+        return { type: "class", parts: [node.value], inverted:false, ignoreCase: node.ignoreCase };
```
```suggestion
        return { type: "class", parts: [node.value], inverted: false, ignoreCase: node.ignoreCase };
```

markw65 replied:
> Need to check how this optimization will affect source map generation

Good point. I think there's one small issue I need to address...

Mingun reviewed (approved):
Need to check how this optimization will affect source map generation. Other than that, all is good

hildjj commented:
I'm going off what I see in the `coverage/lcov-report` directory after having run `npm run build`.  That dir can sometimes be a little confusing because it doesn't get cleaned up every time, and the URLs change if you rerun with a subset of the tests.  Try deleting the directory, then navigating to the correct file from `index.html` when you run a different set of tests.

Here are the three lines that I currently see uncovered in the full test suite: 22, 67, 73.

I don't see breakpoints hit on any of those, but checked to make sure that I have everything set up right by having breakpoints fire on other lines in your file.

hildjj commented:
Also, I'm on the Discord channel if you want to try and figure this out in realtime.

markw65 commented:
I rewrote things a little:
 - some things that asClass was checking for were unnecessary because merge would have already handled them
 - preserve locations when merging nodes
 - only call cleanParts after merging all the alternates for efficiency

And then I updated the tests so that everything is covered.

hildjj commented:
@Mingun, one last thumbs-up, please?

hildjj commented:
@Mingun this one needs a final thumbs-up as well, please.

Mingun commented on lib/compiler/passes/merge-character-classes.js:
```diff
@@ -0,0 +1,159 @@
+            processedRules[node.name] = true;
+            merge(ref);
+          }
+          const cls = asClass(ref);
+          // Return a clone, not the original, because the value we return
+          // may get merged into.
+          return cls && {
+            type: cls.type,
+            parts: cls.parts.slice(),
+            inverted: cls.inverted,
+            ignoreCase: cls.ignoreCase,
+            location: node.location,
+          };
```
Maybe it could be better to request cloned data from `asClass`, because we need to clone only `class` nodes, other nodes already created by us as new objects.

So probably
```javascript
return asClass(ref, true/* return a clone */);
```
here and `asClass(..., false)` otherwise else. Or even make clone at line 118
```javascript
prev = cls;
```
It seems the clone is really needed only there

markw65 replied:
Makes sense.

While we're talking about clones though, it seems to be hard to produce generic clones in this project.

If I say `Object.assign({}, objectToClone)` eslint tells me to use `{ ...objectToClone }`, but if I say `{ ...objectToClone }`, it tells me `Parsing error: Unexpected token ...`.

I don't see why we shouldn't use `...` (it works with node14 and later, and will presumably get fixed by rollup)? Can I change the eslint settings?

hildjj replied:
If you double-check that `...` gets replaced by the rollup step, submit a bug to https://github.com/peggyjs/peggyjs-eslint-config and I'll fix it.  There are a couple of other things I want to make sure are up-to-date while I'm in there.

hildjj replied:
(it's actually the typescript step that is likely doing the replacement, which will be easier to check)

Mingun commented on test/behavior/generated-parser-behavior.spec.js:
```diff
@@ -2677,13 +2677,13 @@ describe("generated parser behavior", () => {
           expect(parser).to.failToParse("d", {
             expected: [
-              { type: "literal", text: "a", ignoreCase: false },
-              { type: "literal", text: "b", ignoreCase: false },
-              { type: "literal", text: "c", ignoreCase: false },
+              { type: "literal", text: "aa", ignoreCase: false },
+              { type: "literal", text: "bb", ignoreCase: false },
+              { type: "literal", text: "cc", ignoreCase: false },
             ],
```
Strictly speaking, we should give parser a chance to not fail by parsing `dd` here and below :)

Mingun commented on test/unit/compiler/passes/merge-character-classes.spec.js:
```diff
@@ -0,0 +1,146 @@
+              parts: [["u", "x"]],
+              ignoreCase: false,
+              inverted: false,
+              location: {
+                start: { line: 3, column: 11 },
+                end: { line: 3, column: 22 },
+              },
```
This location covers
```pegjs
'x' / [u-w]
```
Shouldn't it cover
```pegjs
$('x' / [u-w])
```
?

markw65 replied:
will fix

Mingun commented on test/unit/compiler/passes/merge-character-classes.spec.js:
```diff
@@ -0,0 +1,146 @@
+  it("Merges single character literals, class, and ref_rules", () => {
+    expect(pass).to.changeAST(
+      [
+        "one = three / $('a' / 'd') / [c-f] / [efh] / [c-h] / [c-g]",
+        "two = 'P' / 'P' / three / 'P' / [Q-T]",
+        "three = $('x' / [u-w])",
+        "four = 'a' / [aaa] / 'a'",
```
Because merging also should work when all literal / classes ignores case, I suggest copy the whole test and add `i` suffix here to test `ignoreCase` case

Mingun reviewed (approved):
My approval, except minor nits.

It seems we can slightly optimize by reducing number of unnecessary clones, but benefit also seems to be very low. So can leave this in current state.

We can add tests for `ignoreCase` cases.

Also I suggest to insert `session.emitInfo()` calls when node changed.

markw65 commented:
Anything else to do here?

markw65 commented:
@hildjj @Mingun ping again?
